### DISCUSSION TITLE
Model adds new capacity while historical_new_capacity should be enough

### DISCUSSION THREAD

**Discussion by zsalimian** [iiasa/message_ix#459] [MESSAGE modeling] [answered]
Dear Sir,

I have encountered with a problem in using MESSAGE-ix equations.
I used the GAMS codes for a simple RES.
As you can see this simple model contains 4 years, in which 690 is historical year and 700 is defined to be the first year.
    year_all        /690,700,710,720/
    year /700,710,720/
    first_period(year_all)    /700/
    duration_time(time)=1;
    duration_period(year_all)=10;  
Moreover, I defined two levels, and two technologies. 
    level          /Primary,final/
    tec             /wind_ppl,coal_ppl/
I considered 60-year period for lifetime of each technology.
technical_lifetime(node,'coal_ppl',year)=60;
technical_lifetime(node,'wind_ppl',year)=60;
Furthermore,the coal_ppl has historical_new_capacity(node,'coal_ppl','700')=50;
The demand for electricity is as follow: 
 COMMODITY_USE.fx(node,'electricity','final','700')=1;
 COMMODITY_USE.fx(node, 'electricity' ,'final','710')=1.5;
 COMMODITY_USE.fx(node, 'electricity' ,'final','720')=1.9;
It is expected that the model uses  the following capacity to satisfy the demand   historical_new_capacity(node,'coal_ppl','700') =50 
However, As you see it constructs new capacities for each period, while the lifetime of capacities is 60.
It has made me confused. 

Regards

Sets
    node            world - regions - countries - grid cells
/country/
    commodity       resources - electricity - water - land availability - etc.
/wind,electricity,coal/
    level           levels of the reference energy system or supply chain ( primary - secondary - ... - useful )
/Primary,final/
    tec             technologies
/wind_ppl,coal_ppl/
    mode            modes of operation
/standard/

*    land_type       types of land use
    year_all        years (over entire model horizon)
/690,700,710,720/
    year (year_all) years included in a model instance (for myopic or rolling-horizon optimization)
/700,710,720/
    time            subannual time periods (seasons - days - hours)
/year/

;

* definition of aliases
 Alias(node,location);
*Alias(node,node2);
*Alias(node,node_share);
*Alias(tec,tec2);
*Alias(commodity,commodity2);
*Alias(level,level2);
*Alias(emission,emission2);
 Alias(year_all,vintage);
 Alias(year_all,year_all2);
 Alias(year_all,year_all3);
 Alias(year,year2);
*Alias(year,year3);
 Alias(time,time2);
*Alias(time,time3);
 Alias(time,time_act);
 Alias(time,time_od);
*Alias(mode,mode2);
*Alias(commodity,commodity2);

* category types and mappings
Sets

    type_tec                                types of technologies
/coal_ppl,wind_ppl/

*    cat_tec(type_tec,tec)                   mapping of technologies to respective categories
    inv_tec(tec)                            technologies that have explicit investment and capacity decision variables
/coal_ppl,wind_ppl/

    renewable_tec(tec)                      technologies that use renewable energy potentials
/wind_ppl/
*    addon(tec)                              technologies that are an add-on to other (parent) technologies
*    type_addon                              types of add-on technologies (that can be applied mutually exclusive)
*    cat_addon(type_addon,addon)             mapping of add-on technologies to respective add-on technology types
*    type_year                               types of year aggregations
*/first_model_year/
*    cat_year(type_year,year_all)            mapping of years to respective categories
    type_emission                           types of emission aggregations
/greenhouse gases, pollutants/
*    cat_emission(type_emission,emission)    mapping of emissions to respective categories
*    type_tec_land(type_tec)                 dynamic set whether emissions from land use are included in type_tec
*    balance_equality(commodity,level)       mapping of commodities-level where the supply-demand balance must be maintained with equality
* storage_tec and level_storage are defined at the bottom of this file.
;

*----------------------------------------------------------------------------------------------------------------------*
* assignment and computation of auxiliary parameters                                                                   *
*----------------------------------------------------------------------------------------------------------------------*

* additional sets and parameters created in GAMS to make notation more concise for myopic/rolling-horizon optimization
Sets
    first_period(year_all)           flag for first period in model horizon
/700/
;

Sets
    node            world - regions - countries - grid cells
    commodity       resources - electricity - water - land availability - etc.
    level           levels of the reference energy system or supply chain ( primary - secondary - ... - useful )
    sector          sectors (for integration with MACRO)
    grade           grades of extraction of raw materials
/1/
    tec             technologies
    mode            modes of operation
    emission        greenhouse gases - pollutants - etc.
    land_scenario   scenarios of land use (for land-use model emulator)
    land_type       types of land use
    year_all        years (over entire model horizon)
    year (year_all) years included in a model instance (for myopic or rolling-horizon optimization)
    time            subannual time periods (seasons - days - hours)
    shares          share constraint relations
    relation        generic linear relations
    lvl_spatial     hierarchical levels of spatial resolution
    lvl_temporal    hierarchical levels of temporal resolution
    rating          identifies the 'quality' of the renewable energy potential (bins acc. to Sullivan)
;

* definition of aliases
Alias(node,location);
Alias(node,node2);
Alias(node,node_share);
Alias(tec,tec2);
Alias(commodity,commodity2);
Alias(level,level2);
Alias(emission,emission2);
Alias(year_all,vintage);
Alias(year_all,year_all2);
Alias(year_all,year_all3);
Alias(year,year2);
Alias(year,year3);
Alias(time,time2);
Alias(time,time3);
Alias(time,time_act);
Alias(time,time_od);
Alias(mode,mode2);
Alias(commodity,commodity2);

Sets
*    historical(year_all)             set of periods prior to the start of the model horizon
*    model_horizon(year_all)          set of periods included in the model horizon
*    macro_horizon(year_all)          set of periods included in the MACRO model horizon
    seq_period(year_all,year_all2)    mapping of one period ('year_all') to the next ('year_all2')

*    map_period(year_all,year_all2)    mapping of one period ('year_all') to itself and all subsequent periods ('year_all2')
*    map_first_period(type_year, year_all) mapping of a 'type_year' to the first 'year'
*    first_period(year_all)           flag for first period in model horizon
    last_period(year_all)            flag for last period in model horizon
/720/
*    macro_initial_period(year_all)   flag for period in model horizon in which to initialize model parameters in (period prior to first model period) - used in MACRO
*    macro_base_period(year_all)      flag for base year period in model horizon (period prior to first model period) - used in MACRO
;

seq_period('690','700')=yes;
*seq_period('695','710')=no;
*seq_period('690','720')=no;
seq_period('700','710')=yes;
seq_period('710','720')=yes;
*seq_period('700','720')=no;
*seq_period('710','720')=yes;
*seq_period('720','720')=no;

* category types and mappings
Sets
    level_resource (level)                  subset of 'level' to mark all levels related to make hfossil resources
/Primary/
    level_renewable(level)                  subset of 'level' to mark all levels related to renewable resources
/Primary/

;

Alias(type_tec,type_tec_share);
Alias(type_tec,type_tec_total);

Sets
    map_node(node,location)                     mapping of nodes across hierarchy levels (location is in node)
    map_time(time,time2)                        mapping of time periods across hierarchy levels (time2 is in time)
* map_time_period and map_tec_storage are defined at the bottom of this file.

    map_resource(node,commodity,grade,year_all)  mapping of resources and grades to node over time

    map_ren_grade(node,commodity,grade,year_all) mapping of renewables and grades to node over time
    map_ren_com(node,tec,commodity,year_all)     mapping of technologies to renewable energy source as input
    map_rating(node,tec,commodity,level,rating,year_all) mapping of technologies to ratings bin assignment

    map_commodity(node,commodity,level,year_all,time)    mapping of commodity-level to node and time
    map_stocks(node,commodity,level,year_all)    mapping of commodity-level to node and time

    map_tec(node,tec,year_all)                   mapping of technology to node and years
    map_tec_time(node,tec,year_all,time)         mapping of technology to temporal dissagregation (time)
    map_tec_mode(node,tec,year_all,mode)         mapping of technology to modes
    map_tec_act(node,tec,year_all,mode,time)     mapping of technology to modes AND temporal dissagregation
*    map_tec_addon(tec,type_addon)                mapping of types of add-on technologies to the underlying parent technology

    map_spatial_hierarchy(lvl_spatial,node,node)    mapping of spatial resolution to nodes (last index is 'parent')
    map_temporal_hierarchy(lvl_temporal,time,time)  mapping of temporal resolution to time (last index is 'parent')

    map_shares_commodity_share(shares,node,
        node,type_tec,mode,commodity,level)   mapping for commodity share constraints (numerator)
    map_shares_commodity_total(shares,node,
        node,type_tec,mode,commodity,level)   mapping for commodity share constraints (denominator)

    map_land(node,land_scenario,year_all)            mapping of land-use model emulator scenarios to nodes and years
    map_relation(relation,node,year_all)             mapping of generic (user-defined) relations to nodes and years
;

* additional sets created in GAMS to make notation more concise
Sets
    map_tec_lifetime(node,tec,vintage,year_all)  mapping of technologies to periods within technical lifetime
;

map_resource(node,commodity,grade,year_all)=yes;
*map_resource(node,'oil',grade,year_all)=yes;
*map_resource(node,'wind',grade,year_all)=yes;
*map_resource(node,commodity,grade,year_all)=yes;
map_tec_act(node,tec,year_all,mode,time)=yes;
map_tec_lifetime(node,tec,vintage,year_all)=yes;

*map_tec_lifetime(node,tec,'700','710')=no;
map_tec_time(node,tec,year_all,time)=yes;
map_tec(node,tec,year_all)=yes;
map_ren_com(node,'wind_ppl','wind',year_all)=yes;
map_ren_grade(node,'wind','1',year_all)=yes;
map_time(time,time2)=yes;

Parameters
* general parameters
    interestrate(year_all)         interest rate (to compute discount factor)
    duration_time(time)            duration of one time slice (relative to 1)
    duration_period(year_all)      duration of one multi-year period (in years)
    duration_period_sum(year_all,year_all2)  number of years between two periods ('year_all' must precede 'year_all2')
    df_period(year_all)            cumulative discount factor over period duration
    df_year(year_all)              discount factor of the last year in the period
;

Parameters
* general parameters
*    interestrate(year_all)         interest rate (to compute discount factor)
*    duration_time(time)            duration of one time slice (relative to 1)
*    duration_period(year_all)      duration of one multi-year period (in years)
*    duration_period_sum(year_all,year_all2)  number of years between two periods ('year_all' must precede 'year_all2')
    duration_time_rel(time,time2)  relative duration of subannual time period ('time2' relative to parent 'time')
*    df_period(year_all)            cumulative discount factor over period duration
*    df_year(year_all)              discount factor of the last year in the period
;

Parameters
* technology input-output mapping and costs parameters
    var_cost(node,tec,vintage,year_all,mode,time)       variable costs of operation (per unit of capacity maintained)
    inv_cost(node,tec,year_all)                         investment costs (per unit of new capacity)
    fix_cost(node,tec,vintage,year_all)                 fixed costs per year (per unit of capacity maintained)

* engineering parameters
    construction_time(node,tec,vintage)                     duration of construction of new capacity (in years)
    technical_lifetime(node,tec,vintage)                    maximum technical lifetime (from year of construction)
    capacity_factor(node,tec,vintage,year_all,time)         capacity factor by subannual time slice
    operation_factor(node,tec,vintage,year_all)             yearly total operation factor

;

Parameters
     construction_time_factor(node,tec,year_all) scaling factor to account for construction time of new capacity
     remaining_capacity(node,tec,year_all,year_all) scaling factor to account for remaining capacity in period
     end_of_horizon_factor(node,tec,year_all)    multiplier for value of investment at end of model horizon
     beyond_horizon_lifetime(node,tec,year_all)  remaining technical lifetime at the end of model horizon
     beyond_horizon_factor(node,tec,year_all)    discount factor of remaining lifetime beyond model horizon
;

Parameters

    fixed_new_capacity(node,tec,year_all)               fixed new-built capacity
    fixed_capacity(node,tec,vintage,year_all)           fixed maintained capacity
    fixed_activity(node,tec,vintage,year_all,mode,time) fixed activity level
    input(node,tec,vintage,year_all,mode,node,commodity,level,time,time)  relative share of input per unit of activity
;

Parameters
*from Python
    coal_fraction
    demand_per_year
    historic_demand
    historic_generation
    grid_efficiency
;

Parameter

    elapsed_years(year_all)    elapsed years since the start of the model horizon (not including 'year_all' period)
    remaining_years(year_all)  remaining years until the end of the model horizon (including last period)

;

Parameters
    initial_new_capacity_up(node,tec,year_all)     dynamic upper bound on new capacity (fixed initial term)
    initial_activity_up(node,tec,year_all,time)    dynamic upper bound on activity (fixed initial term)
    initial_new_capacity_lo(node,tec,year_all)     dynamic lower bound on new capacity (fixed initial term)
    initial_activity_lo(node,tec,year_all,time)    dynamic lower bound on activity (fixed initial term)
    growth_new_capacity_up(node,tec,year_all)      dynamic upper bound on new capacity (growth rate)
    growth_activity_up(node,tec,year_all,time)     dynamic upper bound on activity (growth rate)
    growth_new_capacity_lo(node,tec,year_all)      dynamic lower bound on new capacity (growth rate)
    growth_activity_lo(node,tec,year_all,time)     dynamic lower bound on activity (growth rate)
    soft_new_capacity_lo(node,tec,year_all)        soft relaxation of dynamic lower bound on new capacity (growth rate)
    soft_new_capacity_up(node,tec,year_all)        soft relaxation of dynamic upper bound on new capacity (growth rate)
    soft_activity_up(node,tec,year_all,time)       soft relaxation of dynamic upper bound on activity (growth rate)
    soft_activity_lo(node,tec,year_all,time)       soft relaxation of dynamic lower bound on activity (growth rate)
;

Parameters
    historical_new_capacity(node,tec,year_all)           historical new capacity
    historical_activity(node,tec,year_all,mode,time)     historical activity
;

Parameters
    act_to_cap
;

Parameters
* technology input-output mapping and costs parameters
    input(node,tec,vintage,year_all,mode,node,commodity,level,time,time)  relative share of input per unit of activity
    output(node,tec,vintage,year_all,mode,node,commodity,level,time,time) relative share of output per unit of activity
    inv_cost(node,tec,year_all)                         investment costs (per unit of new capacity)
    fix_cost(node,tec,vintage,year_all)                 fixed costs per year (per unit of capacity maintained)
    var_cost(node,tec,vintage,year_all,mode,time)       variable costs of operation (per unit of capacity maintained)
    levelized_cost(node,tec,year_all,time)              levelized costs (per unit of new capacity)

* engineering parameters
    construction_time(node,tec,vintage)                     duration of construction of new capacity (in years)
    technical_lifetime(node,tec,vintage)                    maximum technical lifetime (from year of construction)
    capacity_factor(node,tec,vintage,year_all,time)         capacity factor by subannual time slice
    operation_factor(node,tec,vintage,year_all)             yearly total operation factor
    min_utilization_factor(node,tec,vintage,year_all)       yearly minimum utilization factor
    rating_bin(node,tec,year_all,commodity,level,time,rating) maximum share of technology in commodity use per rating
    reliability_factor(node,tec,year_all,commodity,level,time,rating) reliability of a technology (per rating)
    flexibility_factor(node,tec,vintage,year_all,mode,commodity,level,time,rating) contribution of technologies towards operation flexibility constraint
    renewable_capacity_factor(node,commodity,grade,level,year_all) quality of renewable potential grade (>= 1)
    renewable_potential(node,commodity,grade,level,year_all) size of renewable potential per grade
    emission_factor(node,tec,year_all,year_all,mode,emission) emission intensity of activity
;

Parameter
* resource and commodity parameters
    resource_volume(node,commodity,grade)               volume of resources in-situ at start of the model horizon
    resource_cost(node,commodity,grade,year_all)        extraction costs for resource
    resource_remaining(node,commodity,grade,year_all)   maximum extraction relative to remaining resources (by year)
    bound_extraction_up(node,commodity,grade,year_all)  upper bound on extraction of resources by grade
    commodity_stock(node,commodity,level,year_all)      exogenous (initial) quantity of commodity in stock
    historical_extraction(node,commodity,grade,year_all) historical extraction
;
Parameter
    duration_period_sum(year_all,year_all2) number of years between two periods ('year_all' must precede 'year_all2')
    duration_time_rel(time,time2)         relative duration of subannual time period ('time2' relative to parent 'time')
    elapsed_years(year_all)    elapsed years since the start of the model horizon (not including 'year_all' period)
    remaining_years(year_all)  remaining years until the end of the model horizon (including last period)
    year_order(year_all)       order for members of set 'year_all'
;

Parameter

    bound_new_capacity_up(node,inv_tec,year)
;
bound_new_capacity_up(node,inv_tec,year)=100;
* define order of set 'year_all' (to use as equivalent of ORD operator on the dynamic set year (subset of 'year_all') )
year_order(year_all) = ORD(year_all) ;
resource_remaining(node,commodity,grade,year_all)=150;
resource_cost(node,commodity,grade,year_all)=0.01;
interestrate(year)=0;
*first_period(year_all)=700;
duration_time(time)=1;
inv_cost(node,'coal_ppl',year_all)=100;
*inv_cost(node,'thermal_ppl',year_all)=500;
inv_cost(node,'wind_ppl',year_all)=200;

fix_cost(node,'coal_ppl',vintage,year)=30;
*fix_cost(node,'thermal_ppl',vintage,year)=30;
fix_cost(node,'wind_ppl',vintage,year)=30;

var_cost(node,'coal_ppl',vintage,year_all,mode,time)=30;
*var_cost(node,'thermal_ppl',vintage,year_all,mode,time)=30;
var_cost(node,'wind_ppl',vintage,year_all,mode,time)=0;

duration_period(year_all)=10;
* auxiliary parameters for duration between periods (years) - not including the final period 'year_all2'
duration_period_sum(year_all,year_all2) =
    SUM(year_all3$( ORD(year_all) <= ORD(year_all3) AND ORD(year_all3) < ORD(year_all2) ) , duration_period(year_all3) ) ;
input(node,tec,vintage,year_all,mode,node,commodity,level,time,time)=1;
display duration_period_sum;
input(node,'coal_ppl',vintage,year_all,mode,node,'coal',level,time,time)=1;

*farz minkone kole sal dar madare
operation_factor(node,tec,vintage,year_all)=1;

capacity_factor(node,'coal_ppl',year,year,time2)=1;
*capacity_factor(node,'thermal_ppl',year,year,time2)=1;
capacity_factor(node,'wind_ppl',year,year,time2)=0.36;

technical_lifetime(node,'coal_ppl',year)=60;
*technical_lifetime(node,'thermal_ppl',year)=10;
technical_lifetime(node,'wind_ppl',year)=60;

coal_fraction = 0.5;
demand_per_year = 1;

*1., 1.5, 1.9

historic_demand = 0.85 * demand_per_year;
grid_efficiency = 0.9;
historic_generation = historic_demand / grid_efficiency;
*///zarfiat jadidi ke az ghabl fix mikonin ta sale khasi be madar biad
*///baraye tarife zarfiate mojood karbord dare

historical_new_capacity(node,'coal_ppl','700')=50;
*historical_new_capacity(node,'coal_ppl',year_all)=0$(ord(year_all)>699);
*historical_new_capacity(node,'thermal_ppl',year_all)=0;
historical_new_capacity(node,'wind_ppl',year_all)=0;
*min_utilization_factor(node,inv_tec,vintage,year)=0.5;
renewable_potential(node,commodity,grade,level,year_all)=100;

df_year(year_all) = 1 ;

df_period(year_all) =
    df_year(year_all) * (
* multiply the per-year discount factor by the geometric series of over the duration of the period
        ( ( 1 - POWER( 1 / ( 1 + interestrate(year_all) ), duration_period(year_all) ) )
      / ( 1 - 1 / ( 1 + interestrate(year_all) ) ) )$( interestrate(year_all) )
* if interest rate = 0, multiply by the number of years in that period
        + ( duration_period(year_all) )$( interestrate(year_all) eq 0 ) )
;

remaining_capacity(node,'coal_ppl',vintage,year_all)= 1;
*remaining_capacity(node,'coal_ppl','700','710')= 0;
*remaining_capacity(node,'coal_ppl','700','705')= 0.5;
*remaining_capacity(node,'coal_ppl','710','720')= 1;

*remaining_capacity(node,'coal_ppl',vintage,'710')= 1;
*remaining_capacity(node,'coal_ppl',vintage,'700')= 1;
*remaining_capacity(node,'coal_ppl',vintage,'700')= 1;
*remaining_capacity(node,'coal_ppl',vintage,'720')= 1;
*remaining_capacity(node,'coal_ppl',vintage,'700')= 1;
*remaining_capacity(node,'coal_ppl',vintage,'700')= 1;

remaining_capacity(node,tec,vintage,year_all)$( map_tec_lifetime(node,tec,vintage,year_all)
        AND ( technical_lifetime(node,tec,vintage) - duration_period_sum(vintage,year_all) < duration_period(year_all) )
        AND ( technical_lifetime(node,tec,vintage) - duration_period_sum(vintage,year_all) > 0 ) )
    = ( technical_lifetime(node,tec,vintage) - duration_period_sum(vintage,year_all) ) / duration_period(year_all) ;

display remaining_capacity,technical_lifetime,duration_period_sum;
* compute auxiliary parameters for relative duration of subannual time periods
duration_time_rel(time,time2)$( map_time(time,time2) )
= duration_time(time2) / duration_time(time) ;

renewable_capacity_factor(node,commodity,grade,level_renewable,year)=1;

construction_time(node,inv_tec,year)=1;
* set default construction_time_factor to 1
construction_time_factor(node,inv_tec,year)$( map_tec(node,inv_tec,year) )
 = 1;

* compute the construction_time_factor
construction_time_factor(node,inv_tec,year)$( map_tec(node,inv_tec,year) AND construction_time(node,inv_tec,year) )
 =
     ( 1 + interestrate(year) ) ** construction_time(node,inv_tec,year) ;

display construction_time_factor;
* compute the cumulative discount factor of the technical lifetime remaining beyond the model horizon

***
* auxiliary parameter for duration until the end of the model horizon - including the last period
remaining_years(year_all) = SUM(year_all2$( ORD(year_all) <= ORD(year_all2) ) , duration_period(year_all2) ) ;

* compute the scaling investment parameters for all periods in the model horizon
*year(year_all)$( model_horizon(year_all) ) = yes ;

* compute the technical lifetime remaining beyond the overall model horizon
beyond_horizon_lifetime(node,inv_tec,vintage)$( map_tec(node,inv_tec,vintage) )
 =
    technical_lifetime(node,inv_tec,vintage) - remaining_years(vintage) ;
beyond_horizon_lifetime(node,inv_tec,vintage)$( beyond_horizon_lifetime(node,inv_tec,vintage) < 0 ) = 0 ;

beyond_horizon_factor(node,inv_tec,vintage)$( beyond_horizon_lifetime(node,inv_tec,vintage) )
    = sum(last_period,
* compute the discount factor of the very last year (not period) in the model horizon
        df_year(last_period) * POWER( 1 / ( 1 + interestrate(last_period) ), duration_period(last_period) ) * (
* multiply this by the geometric series of remaining technical lifetime if interestrate of last model period > 0
            (
                ( 1 - POWER( 1 / ( 1 + interestrate(last_period) ), beyond_horizon_lifetime(node,inv_tec,vintage) ) )
                / ( 1 - 1 / ( 1 + interestrate(last_period) ) )
            )$( interestrate(last_period) )
* if interest rate = 0, multiply by remaining technical lifetime
            + ( beyond_horizon_lifetime(node,inv_tec,vintage) )$( interestrate(last_period) eq 0 )
        )
    ) ;

* deterine the parameter end_of_horizon_factor used for scaling investment costs to account for
* technical lifetime beyond the model horizon
end_of_horizon_factor(node,inv_tec,vintage)$( map_tec(node,inv_tec,vintage) )
=
    sum(year_all$( map_tec_lifetime(node,inv_tec,vintage,year_all) )
, df_period(year_all)  )
    / ( sum(year_all$( map_tec_lifetime(node,inv_tec,vintage,year_all) )
,
df_period(year_all) )
        + beyond_horizon_factor(node,inv_tec,vintage) ) ;

resource_volume(node,'coal',grade)=1000;
*resource_volume(node,'oil',grade)=1000;

display historical_new_capacity,duration_time_rel,input,beyond_horizon_factor,end_of_horizon_factor,construction_time_factor;

Variables

    COST_NODAL(node, year_all)                   system costs at the node level over time
* auxiliary variables for finrm-capacity formulation
    COMMODITY_USE(node,commodity,level,year_all) total amount of a commodity & level that was used or consumed
;

Variables
    OBJ objective value of the optimisation problem
;

Positive Variables
* resource production/extraction variable
    EXT(node,commodity,grade,year_all)   extraction of fossil resources
    CAP_NEW(node,tec,year_all)       new capacity by year
    CAP(node,tec,vintage,year_all)   total installed capacity by year
* use of renewable resources
    REN(node, tec, commodity, grade, year_all, time)     activity of renewables specified per renewables grade

    ACT(node,tec,vintage,year_all,mode,time)     activity of technology by mode-year-timeperiod
;

* CAP.fx(node,inv_tec,vintage,first_period)=remaining_capacity(node,inv_tec,vintage,first_period) *duration_period(vintage) * historical_new_capacity(node,inv_tec,vintage);
 COMMODITY_USE.fx(node,'electricity','final','700')=demand_per_year;
 COMMODITY_USE.fx(node,'electricity','final','710')=demand_per_year*1.5;
 COMMODITY_USE.fx(node,'electricity','final','720')=demand_per_year*1.9;

Equations
    OBJECTIVE                       objective value of the optimisation problem
    COST_ACCOUNTING_NODAL           cost accounting at node level over time
    EXTRACTION_EQUIVALENCE          auxiliary equation to simplify the resource extraction formulation
    CAPACITY_MAINTENANCE_HIST       constraint for capacity maintenance  historical installation (built before start of model horizon)
    OPERATION_CONSTRAINT            constraint on maximum yearly operation (scheduled down-time for maintenance)
    CAPACITY_CONSTRAINT             capacity constraint for technology (by sub-annual time slice)
    COMMODITY_USE_LEVEL             aggregate use of commodity by level as defined by total input into technologies
    CAPACITY_MAINTENANCE_NEW        constraint for capacity maintenance of new capacity built in the current period (vintage == year)
    RENEWABLES_POTENTIAL_CONSTRAINT constraint on renewable resource potential
    RENEWABLES_CAPACITY_REQUIREMENT lower bound on required overcapacity when using lower grade potentials
    RENEWABLES_EQUIVALENCE          equation to define the renewables extraction
    RESOURCE_HORIZON
    RESOURCE_CONSTRAINT
    CAPACITY_MAINTENANCE
*    MIN_UTILIZATION_CONSTRAINT
*    NEW_CAPACITY_BOUND_UP
*    NEW_CAPACITY_CONSTRAINT_UP
;

OBJECTIVE..
    OBJ =E= SUM( (node,year), df_period(year) * COST_NODAL(node,year) ) ;

COST_ACCOUNTING_NODAL(node, year)..
    COST_NODAL(node, year) =E=
* resource extraction costs
    SUM((commodity,grade)$( map_resource(node,commodity,grade,year) )
,
         resource_cost(node,commodity,grade,year) * EXT(node,commodity,grade,year) )
    + SUM((tec)$( map_tec(node,tec,year) )
,
            ( inv_cost(node,tec,year) * construction_time_factor(node,tec,year)
                * end_of_horizon_factor(node,tec,year) * CAP_NEW(node,tec,year)
            + SUM(vintage$( map_tec_lifetime(node,tec,vintage,year) )
,
                fix_cost(node,tec,vintage,year) * CAP(node,tec,vintage,year) ) )$( inv_tec(tec) )
            + SUM((vintage,mode,time)$( map_tec_lifetime(node,tec,vintage,year) AND map_tec_act(node,tec,year,mode,time) )
,
                var_cost(node,tec,vintage,year,mode,time) * ACT(node,tec,vintage,year,mode,time) )
            )
;

*///tec az sum dovom hazf shode
EXTRACTION_EQUIVALENCE(node,'coal',year)..
    SUM(grade$( map_resource(node,'coal',grade,year) ),EXT(node,'coal',grade,year) )
    =E= SUM((location,vintage,mode,level_resource,time_act,time_od)$( map_tec_act(node,'coal_ppl',year,mode,time_act) AND  map_tec_lifetime(node,'coal_ppl',vintage,year) ),
        input(location,'coal_ppl',vintage,year,mode,node,'coal',level_resource,time_act,time_od)
        * ACT(location,'coal_ppl',vintage,year,mode,time_act) ) ;

***
RESOURCE_CONSTRAINT(node,commodity,grade,year)$( map_resource(node,commodity,grade,year)
        AND resource_remaining(node,commodity,grade,year) )..
* extraction per year
    EXT(node,commodity,grade,year) =L=
* remaining resources multiplied by remaining-resources-factor
    resource_remaining(node,commodity,grade,year)
    * ( resource_volume(node,commodity,grade)
        - SUM(year2$( year_order(year2) < year_order(year) ),
            duration_period(year2) * EXT(node,commodity,grade,year2) ) ) ;

*************************************************************************************************************************
*
CAPACITY_MAINTENANCE_HIST(node,inv_tec,vintage,first_period)..
    CAP(node,inv_tec,vintage,first_period)
    =L= remaining_capacity(node,inv_tec,vintage,first_period) *
        duration_period(vintage) * historical_new_capacity(node,inv_tec,vintage) ;

**
CAPACITY_MAINTENANCE(node,inv_tec,vintage,year)$( map_tec_lifetime(node,inv_tec,vintage,year)
        AND NOT first_period(year) AND year_order(vintage) < year_order(year))..
    CAP(node,inv_tec,vintage,year)
    =L= remaining_capacity(node,inv_tec,vintage,year) *
        ( SUM(year2$( seq_period(year2,year) ),
              CAP(node,inv_tec,vintage,year2) ) ) ;

******///////////////////
CAPACITY_MAINTENANCE_NEW(node,inv_tec,vintage,vintage)$( map_tec_lifetime(node,inv_tec,vintage,vintage) )..
    CAP(node,inv_tec,vintage,vintage)
    =E= remaining_capacity(node,inv_tec,vintage,vintage)
        * duration_period(vintage) * CAP_NEW(node,inv_tec,vintage) ;
;

**********************************************************************************************************************

***
RESOURCE_HORIZON(node,commodity,grade)$( SUM(year$map_resource(node,commodity,grade,year), 1 ) )..
    SUM(year, duration_period(year) * EXT(node,commodity,grade,year) ) =L= resource_volume(node,commodity,grade) ;

***
RENEWABLES_EQUIVALENCE(node,renewable_tec,commodity,year,time)$(
        map_tec(node,renewable_tec,year) AND map_ren_com(node,renewable_tec,commodity,year)
)..
    SUM(grade$( map_ren_grade(node,'wind',grade,year) )
, REN(node,renewable_tec,commodity,grade,year,time) )
    =E= SUM((location,vintage,mode,level_renewable,time_act)$(map_tec_act(node,renewable_tec,year,mode,time) AND map_tec_lifetime(node,renewable_tec,vintage,year) )
,
        input(location,renewable_tec,vintage,year,mode,node,commodity,level_renewable,time_act,time)
        * ACT(location,renewable_tec,vintage,year,mode,time) ) ;

RENEWABLES_CAPACITY_REQUIREMENT(node,inv_tec,commodity,year)$(SUM( (vintage,mode,time,grade,level_renewable),
            map_tec_lifetime(node,inv_tec,vintage,year) AND map_tec_act(node,inv_tec,year,mode,time) AND map_ren_com(node,inv_tec,commodity,year)
            AND
            renewable_capacity_factor(node,commodity,grade,level_renewable,year) > 0 ))..
    SUM( (vintage,time)$map_ren_com(node,inv_tec,commodity,year)
,
        CAP(node,inv_tec,vintage,year)
        * operation_factor(node,inv_tec,vintage,year)
        * capacity_factor(node,inv_tec,vintage,year,time) )
    =G= SUM((grade,time,level_renewable)$(renewable_capacity_factor(node,commodity,grade,level_renewable,year) > 0),
            REN(node,inv_tec,commodity,grade,year,time)
                 / renewable_capacity_factor(node,commodity,grade,level_renewable,year)) ;

***
RENEWABLES_POTENTIAL_CONSTRAINT(node,commodity,grade,year)$( map_ren_grade(node,commodity,grade,year))
..
    SUM((renewable_tec,time)$( map_ren_com(node,renewable_tec,commodity,year) )
,
        REN(node,renewable_tec,commodity,grade,year,time) )
    =L= SUM(level_renewable, renewable_potential(node,commodity,grade,level_renewable,year) ) ;

OPERATION_CONSTRAINT(node,inv_tec,vintage,year)$( map_tec_lifetime(node,inv_tec,vintage,year) AND operation_factor(node,inv_tec,vintage,year) < 1 )..
    SUM((mode,time)$( map_tec_act(node,inv_tec,year,mode,time) )
, ACT(node,inv_tec,vintage,year,mode,time) ) =L=
        operation_factor(node,inv_tec,vintage,year) * capacity_factor(node,inv_tec,vintage,year,'year')
        * CAP(node,inv_tec,vintage,year) ;

CAPACITY_CONSTRAINT(node,inv_tec,vintage,year,time)$( map_tec_time(node,inv_tec,year,time) AND map_tec_lifetime(node,inv_tec,vintage,year) )..
    SUM(mode$( map_tec_act(node,inv_tec,year,mode,time) )
, ACT(node,inv_tec,vintage,year,mode,time) )
        =L= duration_time(time) * capacity_factor(node,inv_tec,vintage,year,time) * CAP(node,inv_tec,vintage,year) ;

***
COMMODITY_USE_LEVEL(node,'electricity','final',year,time)
flexibility_factor(node,tec,vintage,year,mode,commodity,level,time,rating) ) )
..
    COMMODITY_USE(node,'electricity','final',year)
    =E=
    SUM( (location,tec,vintage,mode,time2)$( map_tec_act(location,tec,year,mode,time2) AND map_tec_lifetime(location,tec,vintage,year) )
,
        input(location,tec,vintage,year,mode,node,'electricity','final',time2,time)
        * duration_time_rel(time,time2)
        * ACT(location,tec,vintage,year,mode,time2) ) ;

 option limrow=100000;
Model MESSAGE_LP / all / ;
Solve MESSAGE_LP using lp minimizing OBJ ;

**Comment by behnam-zakeri** [comment, marked as answer]
Thanks @zsalimian for posting your question. From what you have shared, I see that you have combined several GAMS files of MESSAGEix, from loading data and mapping files to the core formulation and post-processing, into one single file. Furthermore, you have specified your input data hard-coded in GAMS and have changed some parts of the GAMS code to fit your purpose. In principle, we encourage the users for flexible use of the code, and modify it as they wish. However, we do not have enough capacity to read new codes developed by different users and comment on that.
So, if you believe a part of MESSAGEix GAMS formulation is not working as intended in the documentation, please replicate the issue using a simple example using the original code. For example, for this case you showed, please use the original code of MESSAGEix and specify your input data in an interface convenient to you (preferably jupyter notebook), then check the model input and output, e.g., by looking into GDX files, and see if the behavior is what you expect or not. You may be able to do this by making some changes to the Westeros baseline scenario. I hope this helps you find a way to resolve your question.

**Reply by glatterf42** [reply]
I'm marking this as an answer since it does not look like this question is actively pursued from the team and no further input has been given. If this topic is still relevant, please open another discussion/issue.
